Assets/Plugins/Source/Core/Utility/EnumUtility.cs

```diff
@@ -0,0 +1,180 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+namespace PlayEveryWare.EpicOnlineServices.Extensions
+{
+    using System;
+    using System.Collections.Generic;
+    using UnityEngine;
+
+    /// <summary>
+    /// Provides support for alternate string parsing of enum values.
+    /// </summary>
+    internal static class EnumUtility<TEnum> where TEnum : struct, Enum
+    {
+        /// <summary>
+        /// Tries to parse the given string into an enum value. If the string
+        /// provided is either empty or null, the value will be set to the
+        /// indicated defaultValue. If the string can be parsed successfully
+        /// using the System.Enum.TryParse method, that result is returned. If
+        /// that approach fails, a lookup is performed on the provided
+        /// customMapping. If that fails, then the result is set to the
+        /// specified default value, and false is returned.
+        /// </summary>
+        /// <param name="enumValueString">
+        /// The string to parse into an enum value.
+        /// </param>
+        /// <param name="customMappings">
+        /// A mapping of string values to enum types.
+        /// </param>
+        /// <param name="result">
+        /// The result of parsing the given string into an enum value of the
+        /// type specified.
+        /// </param>
+        /// <param name="defaultValue">
+        /// The default enum value to use if the provided string is either null,
+        /// empty, or cannot be parsed into a value.
+        /// </param>
+        /// <returns>
+        /// True if the string was successfully parsed into an enum value, false
+        /// otherwise.
+        /// </returns>
+        public static bool TryParse(
+            string enumValueString, 
+            IDictionary<string, TEnum> customMappings, 
+            out TEnum result, 
+            TEnum defaultValue = default)
+        {
+            result = defaultValue;
+
+            // Allow for an empty or null string to indicate the explicit
+            // default enum value indicated should be set.
+            if (string.IsNullOrEmpty(enumValueString))
+            {
+                return true;
+            }
+
+            // Try to parse the string using the Enum.TryParse method, fall back
+            // on the custom mapping dictionary declared above if that fails.
+            if (!Enum.TryParse(enumValueString, out result))
+            {
+                Debug.LogError($"\"{enumValueString}\" was not recognized as a valid {nameof(TEnum)} value, and parsing failed.");
+                return false;
```

Comment:
Shouldn't it check the custom mappings first before returns false on parsing? 
i.e. shouldn't it be the case that if the enum doesn't directly parse, it should try to look it up in the provided custom mappings before it returns false?

Reply:
Ah, yes. Good catch. I fixed the order of operations.